File: core/src/main/java/com/dmi/perfectreader/util/collection/DuplexBuffer.java
package com.dmi.perfectreader.util.collection;

import java.util.Arrays;

import static com.google.common.base.Preconditions.checkArgument;
import static java.lang.Math.abs;

public class DuplexBuffer<T> {
    private final int maxRelativeIndex;

    private final Object[] items;

    public DuplexBuffer(int maxRelativeIndex) {
        items = new Object[2 * maxRelativeIndex + 1];
        this.maxRelativeIndex = maxRelativeIndex;
    }

    public void set(int relativeIndex, T item) {
        checkArgument(abs(relativeIndex) <= maxRelativeIndex);
        items[arrayIndex(relativeIndex)] = item;
    }

    @SuppressWarnings("unchecked")
    public T get(int relativeIndex) {
        checkArgument(abs(relativeIndex) <= maxRelativeIndex);
        return (T) items[arrayIndex(relativeIndex)];
    }

    public void clear() {
        Arrays.fill(items, null);
    }

    public int maxRelativeIndex() {
        return maxRelativeIndex;
    }

    private int arrayIndex(int relativeIndex) {
        return maxRelativeIndex + relativeIndex;
    }
}

File: util/util-common/src/main/kotlin/com/dmi/util/collection/DuplexBuffer.kt
package com.dmi.util.collection

import com.dmi.util.lang.modPositive
import java.lang.Math.*

class DuplexBuffer<T>(val maxRelativeIndex: Int) {
    @Suppress("UNCHECKED_CAST")
    private val items: Array<T?> = arrayOfNulls<Any>(2 * maxRelativeIndex + 1) as Array<T?>
    private var shift = 0

    operator fun set(relativeIndex: Int, item: T?) {
        require(abs(relativeIndex) <= maxRelativeIndex)
        items[arrayIndex(relativeIndex)] = item
    }

    operator fun get(relativeIndex: Int): T? {
        require(abs(relativeIndex) <= maxRelativeIndex)
        return items[arrayIndex(relativeIndex)]
    }

    fun shift(relativeIndex: Int) {
        shift += relativeIndex

        when {
            relativeIndex > 0 -> {
                for (i in -maxRelativeIndex..min(maxRelativeIndex, -maxRelativeIndex + relativeIndex - 1))
                    this[i] = null
            }
            relativeIndex < 0 -> {
                for (i in max(-maxRelativeIndex, maxRelativeIndex + relativeIndex + 1)..maxRelativeIndex)
                    this[i] = null
            }
        }
    }

    fun clear() = items.fill(null)

    private fun arrayIndex(relativeIndex: Int) =
            (maxRelativeIndex + relativeIndex - shift) modPositive items.size

    override fun toString() =
            (-maxRelativeIndex..maxRelativeIndex)
                    .map { i -> "$i: ${get(i)}" }
                    .joinToString(", ")
}